docs/README.md

```diff
@@ -0,0 +1,142 @@
+
+```
+├── newrelic-infra
+├── newrelic-infra-ctl
+└── newrelic-infra-service
+```
+
+### `newrelic-infra-service`
+
+This binary is the daemon managed by the service manager (`systemd`, `upstart`, `init-v`).
+
+Its only purpose is to get safe runtime reload/restart. This is achieved by signaling its child `newrelic-infra`. 
+
+### `newrelic-infra`
+
+This binary owns the whole agent runtime. It can be triggered in stand-alone mode if reload/restart features are not required.  
+
+### `newrelic-infra-ctl`
+
+This is the CLI control command to communicate with the agent daemon.
+
+## Runtime
+
+There's three different runtime steps:
+
+1. Startup
+2. Main runtime
+3. Shutdown
+
+### Startup
+
+#### Connectivity check
+
+The agent performs an initial network connection check against New Relic endpoints before bootstrapping the rest of the runtime.
+
+In case of failure, the agent retries connecting to New Relic till the limit of attempts and time is reached. 
+
+#### Connect
+
+This step attempts to uniquely identify the agent/box.
+
+As hostnames are prone to collision and change, a fingerprint is used to present some host/cloud information that might be valuable for identification, such as hostname, cloud instance id, etc.
+
+The agent retrieves the fingerprinting data and requests a unique identifier to the New Relic identity endpoint. Errors at this point behave as circuit breaker, blocking any data submission to the platform.
+
+In case of failure, the agent retries connecting to New Relic till the limit of attempts and time is reached. This step is run concurrently so it avoids blocking the runtime. 
+
+
+#### Main runtime
+
+The main runtime workflow addresses data processing and submission.
+
+Codebase differentiates different paths for:
+
+- Metrics and events: these share the same workflow, as non dimensional metrics are represented through events.
+- Inventory: stateful data. Might require disk persistence.
+
+##### Data sources
+
+**Host metrics** are retrieved by embedded **samplers**, for example: `ProcessSampler, StorageSampler, ...`.
+
+**Host inventory** is retrieved by embedded **inventory plugins**, for example: `KernelModulesPlugin, DpkgPlugin...`
+
+Each type of source has different workflow paths.
+
+**External services data** is retrieved using integrations. Integrations are managed by the `integrations` package. There are different integration protocol versions. Each defines a [JSON API](https://docs.newrelic.com/docs/integrations/infrastructure-integrations/get-started/understand-use-data-infrastructure-integrations).
+
+##### Data processing
+
+Metrics/Events:
+
+- Event queue is shared for all the events (agents and integrations).
+- When event-queue reaches 1K events, the agent discard new events. In this case it logs this error message: `Could not queue event: Queue is full..`
+  > We already know this is not optimal and that it should change once we add agent-level rate-limiting.
```

Comment:
I would avoid this kind of internal notes and leave them to either issue trackers or the To-do

Reply:
all this section is internal stuff for contributors to understand what's all about, I don't think there should be more "internal tiers" than distinguishing between user & contributor